src/Compilers/Core/Portable/PEWriter/PooledBlobBuilder.cs

```diff
@@ -38,7 +66,39 @@ protected override BlobBuilder AllocateChunk(int minimalSize)
 
         protected override void FreeChunk()
         {
-            s_chunkPool.Free(this);
+            if (ChunkCapacity != PoolChunkSize)
+            {
+                // The invariant of this builder is that it produces BlobBuilder instances that have a 
+                // ChunkCapacity of at least 1024. Essentially inside AllocateChuck the pool must be able
+                // to mindlessly allocate a BlobBuilder where ChunkCapacity is at least 1024.
+                //
+                // To maintain this the code must verify that the returned BlobBuilder instances have 
+                // a backing array of the appropriate size. This array can shrink in practice through code
+                // like the following: 
+                //
+                //      var builder = PooledBlobBuilder.GetInstance();
+                //      builder.LinkSuffix(new BlobBuilder(256));
+                //      builder.Free(); // calls FreeChunk where ChunkCapacity is 256
+                //
+                // This shouldn't happen much in practice due to convention of how builders are used but
+                // it is a legal use of the APIs.
+                s_chunkPool.ForgetTrackedObject(this);
+            }
+            else
+            {
+                s_chunkPool.Free(this);
+            }
+#if DEBUG
+            IsAlive = false;
```

Comment:
>IsAlive = false; [](http://example.com/codeflow?start=12&length=16)

Is this a race? Someone could get the instance from ```s_chunkPool``` after we called ```s_chunkPool.Free``` but before we got here. Consider moving this assignment to the beginning of the method, and, perhaps use VolatileWrite